File: javascript/code_interview/sorting/quickSort.js
var assert = require('assert')

function quickSort (aList) {
  if (aList.length < 2) { return aList }

  var pivot = aList[0]
  var tail = aList.slice(1)
  var less = quickSort(tail.filter(function (e) { return e < pivot }))
  var equal = quickSort(tail.filter(function (e) { return e === pivot }))
  var greater = quickSort(tail.filter(function (e) { return e > pivot }))

  return [].concat(less, pivot, equal, greater)
}

var t0 = [1, 2, 3]
var e0 = [1, 2, 3]
assert.deepEqual(quickSort(t0), e0)

var t1 = [2, 1, 0]
var e1 = [0, 1, 2]
assert.deepEqual(quickSort(t1), e1)

var t2 = [-1, 93, 32, 1, 0]
var e2 = [-1, 0, 1, 32, 93]
assert.deepEqual(quickSort(t2), e2)

console.log('All tests pass!')

File: python/code_interview/sorting/quicksort.py
def quicksort(arr):
    if len(arr) < 2:
        return arr

    pivot = arr[0]

    lesser = quicksort([x for x in arr[1:] if x < pivot])
    greater = quicksort([x for x in arr[1:] if x > pivot])
    equal = quicksort([x for x in arr[1:] if x == pivot])
    return lesser + [pivot] + equal + greater

t0 = [1, 2, 3, 4]
assert quicksort(t0) == [1, 2, 3, 4]

t1 = [6, 2, 3, 8]
assert quicksort(t1) == [2, 3, 6, 8]

t2 = [1, 5, 3, 3]
assert quicksort(t2) == [1, 3, 3, 5]

print 'All tests pass!'
